### DISCUSSION TITLE
ModuleNotFoundError: No module named 'bleak'

### DISCUSSION THREAD

**Discussion by daxiong2009** [pyinstaller/pyinstaller#8898] [Help] [answered]
i was testing a simple program that connect and read a bluetooth device using bleak package. the program run normally with .py file but fail to run .exe as the bleak can not be imported. Later i added the multiprocessing part, but the issue is still there.  

<img width="493" alt="屏幕截图 2024-11-28 145530" src="https://github.com/user-attachments/assets/d6dc4194-9b08-41cd-8d81-d7caefd14974">

i think mine is the same as  #8395, but could not fix the issue. Please help

**Comment by bwoodsend** [comment]
If you want help then we need code (I can't run a screenshot) and an error message.

**Reply by daxiong2009** [reply]
sorry,  here is the code:
```python
if __name__ == '__main__':
    import multiprocessing
    multiprocessing.freeze_support()


import asyncio
import bleak


address = "04:EE:03:5A:52:1B"
uuid = "00002a58-0000-1000-8000-00805f9b34fb"
runtime = 10

async def main():
    async with bleak.BleakClient(address) as client:
            print(f"Connected to the device: {address}")
            task = asyncio.create_task(client.start_notify(uuid, emg_changed))
            print("Notifications started...")
            await task
            # following two lines of code can be replaced with block comment if want to run device forever
            await asyncio.sleep(runtime)
            await client.stop_notify(uuid)


def emg_changed(handle, data):
    data_convert = int.from_bytes(data, "little", signed=False)
    print(data_convert)


if __name__ == '__main__':
    asyncio.run(main())


```

**Comment by rokm** [comment]
If this is with `onefile` mode and if that screenshot contains all the code there is, then the issue might be that the main process is not waiting for the spawned child process to finish. So it exits right after spawning the child process, which triggers cleanup of the temporary directory by the `onefile` parent, and so the spawned child process might end up without files.

**Reply by daxiong2009** [reply]
yes this screenshot contain all code and is onefile mode. but i call multiprocessing.freeze_support() as recommended, but does not work, could you please to help me to fix this issue?

**Reply by rokm** [reply]
> yes this screenshot contain all code and is onefile mode. but i call multiprocessing.freeze_support() as recommended, but does not work, could you please to help me to fix this issue?

Have you tried waiting for the multiprocessing.Process() that you start to finish? i.e.

```python
p = Process(...)
p.start()
p.join()
```

**Reply by daxiong2009** [reply]
yes, i tried, but same issue

**Reply by rokm** [reply]
Do you get the same issue if you try to freeze the following test program?

```python
import bleak
print("hello")
```

**Reply by daxiong2009** [reply]
this one works

**Reply by rokm** [reply]
Can you post the full error message you are getting with your example, then?

**Reply by daxiong2009** [reply]
this is warn-xx:

This file lists modules PyInstaller was not able to find. This does not
necessarily mean this module is required for running your program. Python and
Python 3rd-party packages include a lot of conditional or optional modules. For
example the module 'ntpath' only exists on Windows, whereas the module
'posixpath' only exists on Posix systems.

Types if import:
* top-level: imported at the top-level - look at these first
* conditional: imported within an if-statement
* delayed: imported within a function
* optional: imported within a try-except-statement

IMPORTANT: Do NOT post this list to the issue-tracker. Use it as a basis for
            tracking down the missing module yourself. Thanks!

missing module named _frozen_importlib_external - imported by importlib._bootstrap (delayed), importlib (optional), importlib.abc (optional), zipimport (top-level)
excluded module named _frozen_importlib - imported by importlib (optional), importlib.abc (optional), zipimport (top-level)
missing module named pwd - imported by posixpath (delayed, conditional, optional), shutil (delayed, optional), tarfile (optional), pathlib (delayed, optional), subprocess (delayed, conditional, optional), netrc (delayed, conditional), getpass (delayed)
missing module named grp - imported by shutil (delayed, optional), tarfile (optional), pathlib (delayed, optional), subprocess (delayed, conditional, optional)
missing module named posix - imported by os (conditional, optional), posixpath (optional), shutil (conditional), importlib._bootstrap_external (conditional)
missing module named resource - imported by posix (top-level)
missing module named _scproxy - imported by urllib.request (conditional)
missing module named termios - imported by getpass (optional)
missing module named multiprocessing.BufferTooShort - imported by multiprocessing (top-level), multiprocessing.connection (top-level)
missing module named multiprocessing.AuthenticationError - imported by multiprocessing (top-level), multiprocessing.connection (top-level)
missing module named _posixshmem - imported by multiprocessing.resource_tracker (conditional), multiprocessing.shared_memory (conditional)
missing module named _posixsubprocess - imported by subprocess (conditional), multiprocessing.util (delayed)
missing module named multiprocessing.get_context - imported by multiprocessing (top-level), multiprocessing.pool (top-level), multiprocessing.managers (top-level), multiprocessing.sharedctypes (top-level)
missing module named multiprocessing.TimeoutError - imported by multiprocessing (top-level), multiprocessing.pool (top-level)
missing module named fcntl - imported by subprocess (optional)
missing module named multiprocessing.set_start_method - imported by multiprocessing (top-level), multiprocessing.spawn (top-level)
missing module named multiprocessing.get_start_method - imported by multiprocessing (top-level), multiprocessing.spawn (top-level)
missing module named pyimod02_importers - imported by D:\project\bleak_test\.venv\Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_pkgutil.py (delayed)
missing module named bleak_winrt - imported by bleak.backends.winrt.scanner (conditional)
missing module named 'bleak_winrt.windows' - imported by bleak.backends.winrt.client (conditional), bleak.backends.winrt.characteristic (conditional), bleak.backends.winrt.descriptor (conditional), bleak.backends.winrt.service (conditional)
missing module named libdispatch - imported by bleak.backends.corebluetooth.CentralManagerDelegate (top-level)
missing module named Foundation - imported by bleak.backends.corebluetooth.scanner (top-level), bleak.backends.corebluetooth.CentralManagerDelegate (top-level), bleak.backends.corebluetooth.utils (top-level), bleak.backends.corebluetooth.client (top-level), bleak.backends.corebluetooth.PeripheralDelegate (top-level)
missing module named CoreBluetooth - imported by bleak.backends.corebluetooth.scanner (top-level), bleak.backends.corebluetooth.CentralManagerDelegate (top-level), bleak.backends.corebluetooth.utils (top-level), bleak.backends.corebluetooth.client (top-level), bleak.backends.corebluetooth.characteristic (top-level), bleak.backends.corebluetooth.descriptor (top-level), bleak.backends.corebluetooth.PeripheralDelegate (top-level), bleak.backends.corebluetooth.service (top-level)
missing module named objc - imported by bleak.backends.corebluetooth (top-level), bleak.backends.corebluetooth.scanner (top-level), bleak.backends.corebluetooth.CentralManagerDelegate (top-level), bleak.backends.corebluetooth.PeripheralDelegate (top-level)
missing module named 'dbus_fast.message' - imported by bleak.backends.bluezdbus.client (top-level), bleak.backends.bluezdbus.utils (top-level), bleak.backends.bluezdbus.signals (top-level)
missing module named 'dbus_fast.constants' - imported by bleak.backends.bluezdbus.client (top-level), bleak.backends.bluezdbus.utils (top-level)
missing module named 'dbus_fast.auth' - imported by bleak.backends.bluezdbus.utils (top-level)
missing module named 'dbus_fast.validators' - imported by bleak.backends.bluezdbus.signals (top-level)
missing module named 'dbus_fast.errors' - imported by bleak.backends.bluezdbus.signals (top-level)
missing module named 'dbus_fast.aio' - imported by bleak.backends.bluezdbus.manager (top-level), bleak.backends.bluezdbus.signals (top-level)
missing module named 'dbus_fast.service' - imported by bleak.backends.bluezdbus.advertisement_monitor (top-level)
missing module named dbus_fast - imported by bleak.backends.bluezdbus.client (top-level), bleak.backends.bluezdbus.manager (top-level), bleak.backends.bluezdbus.scanner (top-level)
missing module named jnius - imported by bleak.backends.p4android.client (top-level), bleak.backends.p4android.defs (top-level), bleak.backends.p4android.utils (top-level), bleak.backends.p4android.scanner (top-level)
missing module named 'android.permissions' - imported by bleak.backends.p4android.scanner (top-level)
missing module named 'android.broadcast' - imported by bleak.backends.p4android.scanner (top-level)
missing module named vms_lib - imported by platform (delayed, optional)
missing module named 'java.lang' - imported by platform (delayed, optional)
missing module named java - imported by platform (delayed)
missing module named _winreg - imported by platform (delayed, optional)
missing module named 'dbus_fast.signature' - imported by bleak.backends.bluezdbus.client (top-level)
missing module named android - imported by bleak.backends.p4android.client (top-level)
missing module named async_timeout - imported by bleak (conditional), bleak.backends.bluezdbus.client (conditional), bleak.backends.p4android.scanner (conditional), bleak.backends.corebluetooth.CentralManagerDelegate (conditional), bleak.backends.corebluetooth.PeripheralDelegate (conditional), bleak.backends.winrt.client (conditional)
missing module named asyncio.DefaultEventLoopPolicy - imported by asyncio (delayed, conditional), asyncio.events (delayed, conditional)
missing module named multiprocessing.freeze_support - imported by multiprocessing (top-level), D:\project\bleak_test\main.py (top-level)
missing module named multiprocessing.Process - imported by multiprocessing (top-level), D:\project\bleak_test\main.py (top-level)

**Reply by rokm** [reply]
What is the full error message you are getting when you run your program?

**Reply by daxiong2009** [reply]
where to get full error message? sorry i am new to pyinstaller

**Reply by rokm** [reply]
> where to get full error message? sorry i am new to pyinstaller

You are saying that the `bleak` module cannot be imported (`ModuleNotFoundError: No module named 'bleak'`) - where are you seeing this error? 

Presumably when you run your program? In that case, you should get full traceback as well, either in console window (if your program is console-enabled and you run it in command prompt) or in unhandled exception error dialog (if you are using noconsole/windowed build).

**Reply by rokm** [reply]
Note that the warnings in `warn-xx` file are not necessarily issues (especially for conditional imports), and you can ignore them if the generated executable runs without errors.

**Reply by daxiong2009** [reply]
this should be the full error message. i paste part of them here , i also attach the full log txt file down here

```
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9D8F0>, <winrt._winrt...0023DEAD9DD30>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9D8F0>, <winrt._winrt...0023DEAD9DD30>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9D930>, <winrt._winrt...0023DEAD9DD70>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9D930>, <winrt._winrt...0023DEAD9DD70>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DCB0>, <winrt._winrt...0023DEAD9DCD0>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DCB0>, <winrt._winrt...0023DEAD9DCD0>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DC10>, <winrt._winrt...0023DEAD9D990>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DC10>, <winrt._winrt...0023DEAD9D990>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DDD0>, <winrt._winrt...0023DEAD9DC90>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DDD0>, <winrt._winrt...0023DEAD9DC90>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DE50>, <winrt._winrt...0023DEAD9DE30>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DE50>, <winrt._winrt...0023DEAD9DE30>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DED0>, <winrt._winrt...0023DEAD9DEB0>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DED0>, <winrt._winrt...0023DEAD9DEB0>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DF50>, <winrt._winrt...0023DEAD9DF30>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DF50>, <winrt._winrt...0023DEAD9DF30>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DFD0>, <winrt._winrt...0023DEAD9DFB0>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9DFD0>, <winrt._winrt...0023DEAD9DFB0>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E050>, <winrt._winrt...0023DEAD9E030>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E050>, <winrt._winrt...0023DEAD9E030>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E0D0>, <winrt._winrt...0023DEAD9E0B0>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E0D0>, <winrt._winrt...0023DEAD9E0B0>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E150>, <winrt._winrt...0023DEAD9E130>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E150>, <winrt._winrt...0023DEAD9E130>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E1D0>, <winrt._winrt...0023DEAD9E1B0>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E1D0>, <winrt._winrt...0023DEAD9E1B0>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
Exception in callback BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E250>, <winrt._winrt...0023DEAD9E230>)
handle: <Handle BleakScannerWinRT._received_handler(<winrt._winrt...0023DEAD9E250>, <winrt._winrt...0023DEAD9E230>)>
Traceback (most recent call last):
  File "asyncio\events.py", line 88, in _run
  File "bleak\backends\winrt\scanner.py", line 139, in _received_handler
ModuleNotFoundError: No module named 'winrt.windows.foundation.collections'
```


[exe.log.txt](https://github.com/user-attachments/files/17960078/exe.log.txt)

**Reply by rokm** [reply, marked as answer]
Any difference if you add `--hiddenimport winrt.windows.foundation.collections` to your PyInstaller command (or if you are building with .spec, if you add that module to `hiddenimports` list)?

**Reply by rokm** [reply]
If you subsequently get errors for other modules from `winrt.windows.foundation`, you can try using `--collect-submodules winrt.windows.foundation` instead. EDIT: nevermind, I see that it's the only submodule there.

**Reply by daxiong2009** [reply]
> Any difference if you add `--hiddenimport winrt.windows.foundation.collections` to your PyInstaller command (or if you are building with .spec, if you add that module to `hiddenimports` list)

that works! thanks a lot!
